### PR TITLE
admin: routes endpoint

### LINKED ISSUE #811: add ability to query route table using admin interface
when i use RDS to create the route table dynamic， but i want to know the route table in envoy runtime, how can i do that?

### PR DESCRIPTION
**Pull request by junr03**
add /routes admin endpoint in order to dump dynamically loaded route tables via RDS. The endpoint outputs the information in valid JSON, which can then be consumed an analyzed. Note that static route tables are not dumped by the endpoint as they are already dumped by the /config endpoint.

fixes #811

### CODE REVIEW

File: include/envoy/server/admin.h [hunk 1]
```diff
@@ -11,6 +11,9 @@
 namespace Envoy {
 namespace Server {
 
+#define MAKE_HANDLER(X)                                                                            \
+  [this](const std::string& url, Buffer::Instance& data) -> Http::Code { return X(url, data); }
```
Comment (htuch) [inline include/envoy/server/admin.h]: Does this need to be a macro, or could it be done as a template?

File: source/common/router/rds_impl.cc [hunk 2]
```diff
@@ -174,5 +183,18 @@ Router::RouteConfigProviderSharedPtr RouteConfigProviderManagerImpl::getRouteCon
   return new_provider;
 };
 
+Http::Code RouteConfigProviderManagerImpl::handlerRoutes(const std::string&,
+                                                         Buffer::Instance& response) {
+  for (auto provider : routeConfigProviders()) {
+    response.add(fmt::format(
+        "route_config_name: {}\n",
+        std::static_pointer_cast<RdsRouteConfigProviderImpl>(provider)->route_config_name_));
```
Comment (htuch) [inline source/common/router/rds_impl.cc]: Seems a `configAsString()` message on the `routeConfigProviderr` woudl avoid this casting.

Comment (junr03) [inline source/common/router/rds_impl.cc]: we would still need it for the route name and the cluster name, which only make sense for Rds.

File: include/envoy/server/admin.h (hunk 1 above)
Comment (junr03) [inline include/envoy/server/admin.h]: I have to admit I am not fully aware about how to make that call. Open to learn how you think about it :)

File: source/common/router/rds_impl.cc [hunk 3]
```diff
@@ -174,5 +184,44 @@ Router::RouteConfigProviderSharedPtr RouteConfigProviderManagerImpl::getRouteCon
   return new_provider;
 };
 
+void RouteConfigProviderManagerImpl::addRouteInfo(RouteConfigProviderSharedPtr provider,
+                                                  Buffer::Instance& response) {
+  response.add(fmt::format(
+      "route_config_name: {}\n",
+      std::static_pointer_cast<RdsRouteConfigProviderImpl>(provider)->route_config_name_));
```
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: Per @htuch you shouldn't be using static_cast here (and note even if you did you should be using dynamic_cast). Refactor interfaces as needed to make it work properly.

File: include/envoy/server/admin.h (hunk 1 above)
Comment (mattklein123) [inline include/envoy/server/admin.h]: I don't remember why I originally made this a macro. I think it would be more typing to make it a template and/or factory static method that creates an std::function. Either way you should rename to MAKE_ADMIN_HANDLER if you are moving out to this scope.

File: source/common/router/rds_impl.cc [hunk 4]
```diff
@@ -122,9 +124,17 @@ void RdsRouteConfigProviderImpl::registerInitTarget(Init::Manager& init_manager)
...
+    const LocalInfo::LocalInfo& local_info, ThreadLocal::SlotAllocator& tls, Server::Admin& admin)
     : runtime_(runtime), dispatcher_(dispatcher), random_(random), local_info_(local_info),
-      tls_(tls) {}
+      tls_(tls), admin_(admin) {
+  admin_.addHandler("/routes", "print out currently loaded dynamic route tables",
+                    MAKE_HANDLER(RouteConfigProviderManagerImpl::handlerRoutes), true);
+}
+
+RouteConfigProviderManagerImpl::~RouteConfigProviderManagerImpl() {
+  // delete routes handler
```
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: nit: del obvious comment.

File: source/server/config_validation/admin.h [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace Envoy {
+namespace Server {
+
+/**
+ * Config-validation-only implementation Server::Admin
+ */
+class ValidationAdmin : public Admin {
+public:
+  ValidationAdmin() {}
```
Comment (mattklein123) [inline source/server/config_validation/admin.h]: del empty constructor/destructor.

File: source/common/router/rds_impl.cc (hunk 4 above)
Comment (junr03) [inline source/common/router/rds_impl.cc]: oops yeah, that was for me, forgot to delete.

File: source/common/router/rds_impl.cc (hunk 3 above)
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: Note, one way of doing this is to have RdsRouteConfigProvider interface, and only expose this for iteration. Another way is to also allow printing static route tables and differentiate them somehow.

File: source/common/router/rds_impl.cc [hunk 6]
```diff
@@ -174,5 +184,44 @@ Router::RouteConfigProviderSharedPtr RouteConfigProviderManagerImpl::getRouteCon
...
+  if (it != query_params.end()) {
+    for (auto provider : routeConfigProviders()) {
+      if (std::static_pointer_cast<RdsRouteConfigProviderImpl>(provider)->route_config_name_ ==
+          it->second) {
+        addRouteInfo(provider, response);
+      }
+    }
+    return Http::Code::OK;
+  } else {
+    response.add("usage: /routes? (dump all dynamic route tables)\n");
```
Comment (RomanDzhabarov) [inline source/common/router/rds_impl.cc]: nit: /routes no '?'

File: source/common/router/rds_impl.cc [hunk 7]
```diff
@@ -174,5 +184,44 @@ Router::RouteConfigProviderSharedPtr RouteConfigProviderManagerImpl::getRouteCon
...
+    response.add("usage: /routes?route_config_name=<name> (dump all dynamic route tables with the "
```
Comment (RomanDzhabarov) [inline source/common/router/rds_impl.cc]: with the < name > ?

File: source/server/config_validation/admin.h [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  ~ValidationAdmin() {}
+
+  bool addHandler(const std::string&, const std::string&, HandlerCb, bool) override;
```
Comment (RomanDzhabarov) [inline source/server/config_validation/admin.h]: put comments :)

File: include/envoy/router/route_config_provider_manager.h [hunk 9]
```diff
@@ -56,7 +56,7 @@ class ServerRouteConfigProviderManager : public RouteConfigProviderManager {
    * @return std::vector<Router::RouteConfigProviderSharedPtr> a list of all the
```
Comment (mattklein123) [inline include/envoy/router/route_config_provider_manager.h]: type needs updating.

File: include/envoy/router/route_config_provider_manager.h [hunk 10]
```diff
@@ -56,7 +56,7 @@ class ServerRouteConfigProviderManager : public RouteConfigProviderManager {
...
    * RouteConfigProviders currently loaded.
    */
-  virtual std::vector<RouteConfigProviderSharedPtr> routeConfigProviders() PURE;
+  virtual std::vector<RdsRouteConfigProviderSharedPtr> routeConfigProviders() PURE;
```
Comment (mattklein123) [inline include/envoy/router/route_config_provider_manager.h]: I would probably rename this to rdsRouteConfigProviders() and update comment to make it clear that static providers are not included.

File: include/envoy/server/admin.h [hunk 11]
```diff
@@ -11,6 +11,9 @@
 namespace Envoy {
 namespace Server {
 
+#define MAKE_ADMIN_HANDLER(X)                                                                      \
```
Comment (mattklein123) [inline include/envoy/server/admin.h]: Since this is now global can you add a comment on what this does.

File: include/envoy/router/rds.h [hunk 12]
```diff
@@ -23,7 +23,21 @@ class RouteConfigProvider {
...
 
+/**
+ * A provider for dynamic route configurations.
+ */
+class RdsRouteConfigProvider : public RouteConfigProvider {
+public:
+  virtual ~RdsRouteConfigProvider() {}
+
+  /**
+   * @return string the loaded route table.
```
Comment (mattklein123) [inline include/envoy/router/rds.h]: `@return std::string`

File: include/envoy/server/filter_config.h [hunk 13]
```diff
@@ -116,6 +117,11 @@ class FactoryContext {
    *         used to allow runtime lockless updates to configuration, etc. across multiple threads.
    */
   virtual ThreadLocal::SlotAllocator& threadLocal() PURE;
+
+  /**
+   * @return Admin the server's Global admin HTTP endpoint.
```
Comment (mattklein123) [inline include/envoy/server/filter_config.h]: @return Server::Admin&

File: source/common/router/rds_impl.cc [hunk 14]
```diff
@@ -122,19 +124,26 @@ void RdsRouteConfigProviderImpl::registerInitTarget(Init::Manager& init_manager)
 
 RouteConfigProviderManagerImpl::RouteConfigProviderManagerImpl(
     Runtime::Loader& runtime, Event::Dispatcher& dispatcher, Runtime::RandomGenerator& random,
-    const LocalInfo::LocalInfo& local_info, ThreadLocal::SlotAllocator& tls)
+    const LocalInfo::LocalInfo& local_info, ThreadLocal::SlotAllocator& tls, Server::Admin& admin)
     : runtime_(runtime), dispatcher_(dispatcher), random_(random), local_info_(local_info),
-      tls_(tls) {}
+      tls_(tls), admin_(admin) {
+  admin_.addHandler("/routes", "print out currently loaded dynamic route tables",
```
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: "... HTTP route tables"

File: source/common/router/rds_impl.cc (hunk 3 above)
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: I still see static casts here?

**Comment by junr03** [comment]
@mattklein123 yeah, not done updating yet, I was just pushing at a break point where tests were passing before carrying forward. Some of your comments are already addressed in the upcoming commit. I'll tag you when I am done.

**Comment by junr03** [comment]
@mattklein123 updated.

File: source/common/router/rds_impl.cc [hunk 15]
```diff
@@ -43,6 +43,7 @@ RdsRouteConfigProviderImpl::RdsRouteConfigProviderImpl(
     const std::string& stat_prefix, ThreadLocal::SlotAllocator& tls,
     RouteConfigProviderManagerImpl& route_config_provider_manager)
     : runtime_(runtime), cm_(cm), tls_(tls.allocateSlot()),
+      cluster_name_(rds.config_source().api_config_source().cluster_name()[0]),
```
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: In the future I think theoretically we can have more than a single cluster here, so this isn't going to make sense. Please add a TODO for @htuch to take a look at this call site. I think this is fine for now.

File: source/common/router/rds_impl.cc [hunk 16]
```diff
@@ -174,5 +183,42 @@ Router::RouteConfigProviderSharedPtr RouteConfigProviderManagerImpl::getRouteCon
...
+  response.add("config dump:\n");
+  response.add(fmt::format("{}\n", provider->configAsString()));
+}
+
+Http::Code RouteConfigProviderManagerImpl::handlerRoutes(const std::string& url,
+                                                         Buffer::Instance& response) {
+  Http::Utility::QueryParams query_params = Http::Utility::parseQueryString(url);
+  // If there are no query params, print out all the configured route tables.
+  if (query_params.size() == 0) {
+    for (auto provider : rdsRouteConfigProviders()) {
```
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: const auto&. Same below.

File: source/common/router/rds_impl.cc [hunk 17]
```diff
@@ -174,5 +183,42 @@ Router::RouteConfigProviderSharedPtr RouteConfigProviderManagerImpl::getRouteCon
...
+      addRouteInfo(provider, response);
+    }
+    return Http::Code::OK;
+  }
+
+  // If there are query params, make sure it is only the route_config_name param.
+  auto it = query_params.find("route_config_name");
```
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: const auto

File: include/envoy/router/rds.h [hunk 18]
```diff
@@ -23,7 +23,32 @@ class RouteConfigProvider {
...
+
+  /**
+   * @return std::string the loaded route table.
+   */
+  virtual std::string configAsString() PURE;
+
+  /**
+   * @return std::string the name of the configured route table.
+   */
+  virtual std::string routeConfigName() PURE;
```
Comment (mattklein123) [inline include/envoy/router/rds.h]: const std::string&

File: include/envoy/router/rds.h [hunk 19]
```diff
@@ -23,7 +23,32 @@ class RouteConfigProvider {
...
+
+  /**
+   * @return std::string the name of the cluster the RdsRouteConfigProvider is issuing RDS requests
```
Comment (mattklein123) [inline include/envoy/router/rds.h]: const std::string&

File: source/common/router/rds_impl.h [hunk 20]
```diff
@@ -146,12 +155,30 @@ class RouteConfigProviderManagerImpl : public ServerRouteConfigProviderManager,
                                                       Init::Manager& init_manager) override;
 
 private:
-  std::unordered_map<std::string, std::weak_ptr<RouteConfigProvider>> route_config_providers_;
+  /**
+   * Add the RdsRouteConfigProvider information to the buffer.
+   * @param provider supplies the Provider to extract information from.
+   * @param response supplies the buffer to fill with information.
+   */
+  void addRouteInfo(RdsRouteConfigProviderSharedPtr provider, Buffer::Instance& response);
```
Comment (mattklein123) [inline source/common/router/rds_impl.h]: pass `const RdsRouteConfigProvider&`

File: test/common/router/rds_impl_test.cc [hunk 21]
```diff
@@ -36,9 +41,14 @@ parseHttpConnectionManagerFromJson(const std::string& json_string) {
   return http_connection_manager;
 }
 
-class RdsImplTest : public testing::Test {
+class RdsImplTest : public testing::TestWithParam<Network::Address::IpVersion> {
 public:
-  RdsImplTest() : request_(&cm_.async_client_) {}
+  // TODO(junr03): Switch to mocks and do not bind to a real port.
```
Comment (mattklein123) [inline test/common/router/rds_impl_test.cc]: I would rewrite these tests to use a mock admin here, not a real one. Then you can undo most/all of these changes. You can verify with mocks adding/removing the handler, and also store the handler to call via test.

File: test/common/router/rds_impl_test.cc [hunk 22]
```diff
@@ -253,6 +290,36 @@ TEST_F(RdsImplTest, Basic) {
                        ->routeEntry()
                        ->clusterName());
 
+  // Test Admin /routes handler
+  EXPECT_EQ(Http::Code::OK, admin_.runCallback("/routes", data));
+  EXPECT_EQ("route_config_name: foo_route_config\ncluster_name: foo_cluster\nconfig dump:\nname: "
```
Comment (mattklein123) [inline test/common/router/rds_impl_test.cc]: This test is incredibly difficult to read. Can you redo this using string literals to make it more readable. For stuff above you might consider some kind of utility function that creates a generic response or something.

File: include/envoy/router/rds.h (hunk 18 above)
Comment (htuch) [inline include/envoy/router/rds.h]: Should some of these be `std::string&`? This is just a nit, it might make sense for the implementation to have the freedom to not maintain a const string here.

File: source/common/router/rds_impl.cc [hunk 23]
```diff
@@ -174,5 +183,42 @@ Router::RouteConfigProviderSharedPtr RouteConfigProviderManagerImpl::getRouteCon
   return new_provider;
 };
 
+void RouteConfigProviderManagerImpl::addRouteInfo(RdsRouteConfigProviderSharedPtr provider,
+                                                  Buffer::Instance& response) {
+  response.add(fmt::format("route_config_name: {}\n", provider->routeConfigName()));
+  response.add(fmt::format("cluster_name: {}\n", provider->clusterName()));
+  response.add("config dump:\n");
+  response.add(fmt::format("{}\n", provider->configAsString()));
```
Comment (htuch) [inline source/common/router/rds_impl.cc]: To what extent do we want to make the results from the various handlers easy to parse? Maybe not for this PR, but it seems to make sense to export out the handlers as JSON, since you don't want to have to scrape manually if writing automation that uses the handlers.

File: source/common/router/rds_impl.h [hunk 24]
```diff
@@ -146,12 +155,30 @@ class RouteConfigProviderManagerImpl : public ServerRouteConfigProviderManager,
...
+
+  /**
+   * The handler used in the Admin /routes endpoint. This handler is used to
+   * populate the response Buffer::Instance with information about the currently
+   * loaded dynamic HTTP Route Tables.
+   * @param url supplies the url sent to the Admin endpoint.
+   * @param response supplies the buffer to fill with information.
```
Comment (htuch) [inline source/common/router/rds_impl.h]: Nit: @return

File: source/common/router/rds_impl.cc (hunk 23 above)
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: +1 to just making the output JSON from the start. It should be relatively easy to do. I think the proto can be converted to JSON also @htuch ?

Comment (htuch) [inline source/common/router/rds_impl.cc]: Yep, see `Protobuf::util::MessageToJsonString()`.

**Comment by junr03** [comment]
@mattklein123 @htuch updated.

File: source/common/router/rds_impl.cc [hunk 25]
```diff
@@ -120,21 +123,34 @@ void RdsRouteConfigProviderImpl::registerInitTarget(Init::Manager& init_manager)
   init_manager.registerTarget(*this);
 }
 
+std::string RdsRouteConfigProviderImpl::configAsString() const {
+  std::string json_string;
+  Protobuf::util::MessageToJsonString(route_config_proto_, &json_string);
```
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: See my comment here: https://github.com/lyft/envoy/pull/1517#discussion_r134627775. Can we set the same option? Potentially by just moving that option into the MessageToJsonString function? cc @mrice32 @htuch

File: include/envoy/router/rds.h [hunk 26]
```diff
@@ -23,7 +23,32 @@ class RouteConfigProvider {
...
+ * A provider for dynamic route configurations.
+ */
+class RdsRouteConfigProvider : public RouteConfigProvider {
+public:
+  virtual ~RdsRouteConfigProvider() {}
+
+  /**
+   * @return std::string the loaded route table.
+   */
+  virtual std::string configAsString() const PURE;
```
Comment (mattklein123) [inline include/envoy/router/rds.h]: nit: I would call this configAsJson() now.

File: source/common/router/rds_impl.cc (hunk 25 above)
Comment (junr03) [inline source/common/router/rds_impl.cc]: made a utility function to stop adding this options at all uses. Changed the other call site to use it as well.

File: source/common/protobuf/utility.cc [hunk 27]
```diff
@@ -40,7 +40,11 @@ Json::ObjectSharedPtr WktUtil::getJsonObjectFromStruct(const Protobuf::Struct& m
   const auto status = Protobuf::util::MessageToJsonString(message, &json, json_options);
   // This should always succeed unless something crash-worthy such as out-of-memory.
   RELEASE_ASSERT(status.ok());
-  return Json::Factory::loadFromString(json);
+  return json;
+}
+
+Json::ObjectSharedPtr WktUtil::getJsonObjectFromStruct(const Protobuf::Struct& message) {
```
Comment (htuch) [inline source/common/protobuf/utility.cc]: Nit: since a `Protobuf::Struct` is a `Protobuf::Message`, this function should just be renamed `getJsonObjectFromMessage` (and corresponding arg change to `Message`, move it to `MessageUtil`).

File: source/common/router/rds_impl.cc [hunk 28]
```diff
@@ -174,5 +185,45 @@ Router::RouteConfigProviderSharedPtr RouteConfigProviderManagerImpl::getRouteCon
...
+  const auto it = query_params.find("route_config_name");
+  if (query_params.size() == 1 && it != query_params.end()) {
+    for (const auto& provider : rdsRouteConfigProviders()) {
+      if (provider->routeConfigName() == it->second) {
+        addRouteInfo(*provider, response);
+      }
+    }
+    return Http::Code::OK;
+  }
+  response.add("{\n");
```
Comment (htuch) [inline source/common/router/rds_impl.cc]: Here's an idea that is out-of-scope for this PR, but might be worthy of a TODO or issue filing. How about we define all the handler output formats as protos in `envoy-api`, and just use the JSON transcoding here? This would avoid having to build JSON object by hand and provide the general protoey goodness.

File: source/common/router/rds_impl.cc (hunk 15 above)
Comment (htuch) [inline source/common/router/rds_impl.cc]: Yep, TODO for me works, also tag with https://github.com/lyft/envoy/issues/1170.

File: source/common/router/rds_impl.cc (hunk 28 above)
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: I actually had same thought on this last night. +1. Can you add TODO and also file issue?

Comment (junr03) [inline source/common/router/rds_impl.cc]: +2 while I was hand writing JSON yesterday I thought "isn't this what we are shying away from?".

**Comment by mattklein123** [comment]
LGTM after adding TODO/issue and merging master.

**Comment by junr03** [comment]
@mattklein123 green :). 

btw I am not sure if the mac os hooks are set up correctly. If I click on the details for the failure about I get to the master build, not my build.

**Comment by mattklein123** [comment]
@zuercher ^ FYI on the OSX build failure.

File: include/envoy/server/filter_config.h [hunk 29]
```diff
@@ -120,6 +121,11 @@ class FactoryContext {
    *         used to allow runtime lockless updates to configuration, etc. across multiple threads.
    */
   virtual ThreadLocal::SlotAllocator& threadLocal() PURE;
+
+  /**
+   * @return Server::Admin& the server's Global admin HTTP endpoint.
```
Comment (mattklein123) [inline include/envoy/server/filter_config.h]: nit: s/Global/global

File: test/common/router/rds_impl_test.cc [hunk 30]
```diff
@@ -36,10 +40,19 @@ parseHttpConnectionManagerFromJson(const std::string& json_string) {
   return http_connection_manager;
 }
 
-class RdsImplTest : public testing::Test {
+class RdsImplTest : public testing::TestWithParam<Network::Address::IpVersion> {
```
Comment (mattklein123) [inline test/common/router/rds_impl_test.cc]: you can revert the change on this line.

**Comment by zuercher** [comment]
@junr03 it only looks like master because of the way we trick circle in to building PRs from the upstream repo. You should see you SHA in the build parameters tab.

**Comment by junr03** [comment]
hmmm ok. In that case I am a little stumped why the last commit in this PR caused the failure I saw in MacOS. I might get my mac os env going for curiosity.

**Comment by zuercher** [comment]
There seems to be a sporadic failure in connection_impl_test's BindTest, which may well be unrelated to your changes. I spent some time making a PR for backward-cpp so we can get an actual stack trace in OS X land. Will try to fix the test itself tomorrow.
